File: ios/BlakJaks/Features/Social/ViewModels/SocialViewModel.swift
import Foundation

// MARK: - SocialViewModel

@MainActor
final class SocialViewModel: ObservableObject {

    // MARK: - Published Properties

    @Published var channels: [Channel] = []
    @Published var selectedChannel: Channel? = nil
    @Published var messages: [ChatMessage] = []
    @Published var isLoadingMessages = false
    @Published var isLoadingChannels = false
    @Published var error: Error?
    @Published var currentLiveStream: LiveStream? = MockLiveStream.live
    @Published var isRateLimited = false
    @Published var rateLimitRemainingSeconds: Int = 0
    @Published var newMessageCount = 0
    @Published var pinnedMessage: ChatMessage? = nil
    @Published var draftMessage = ""

    // MARK: - Private Properties

    private var lastSentAt: Date? = nil
    private var rateLimitTimer: Timer? = nil
    private let rateLimitDuration: TimeInterval = 5.0
    private let apiClient: APIClientProtocol
    let currentUserId: Int = 1
    let currentUserTier: String = "Standard"

    // MARK: - Init

    init(apiClient: APIClientProtocol = MockAPIClient()) {
        self.apiClient = apiClient
    }

    // MARK: - Channel Methods

    func loadChannels() async {
        isLoadingChannels = true
        defer { isLoadingChannels = false }
        do {
            channels = try await apiClient.getChannels()
        } catch {
            self.error = error
        }
    }

    func selectChannel(_ channel: Channel) async {
        selectedChannel = channel
        messages = []
        newMessageCount = 0
        await loadMessages()
    }

    func loadMessages(before: Int? = nil) async {
        guard let channelId = selectedChannel?.id else { return }
        isLoadingMessages = true
        defer { isLoadingMessages = false }
        do {
            let fetched = try await apiClient.getMessages(channelId: channelId, limit: 50, before: before)
            if before != nil {
                // Prepend older messages
                messages = fetched + messages
            } else {
                messages = fetched
            }
        } catch {
            self.error = error
        }
    }

    // MARK: - Send Message

    func sendMessage() async {
        let trimmed = draftMessage.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        guard trimmed.count <= 500 else { return }
        guard !isRateLimited else { return }
        guard let channelId = selectedChannel?.id else { return }

        do {
            let sent = try await apiClient.sendMessage(channelId: channelId, content: trimmed, mediaType: nil)
            messages.append(sent)
            draftMessage = ""
            lastSentAt = Date()
            if currentUserTier.lowercased() == "standard" {
                startRateLimitCooldown()
            }
        } catch {
            self.error = error
        }
    }

    // MARK: - Rate Limit Cooldown

    private func startRateLimitCooldown() {
        isRateLimited = true
        rateLimitRemainingSeconds = 5

        rateLimitTimer?.invalidate()
        rateLimitTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] timer in
            guard let self else {
                timer.invalidate()
                return
            }
            Task { @MainActor in
                if self.rateLimitRemainingSeconds > 0 {
                    self.rateLimitRemainingSeconds -= 1
                }
                if self.rateLimitRemainingSeconds <= 0 {
                    self.isRateLimited = false
                    timer.invalidate()
                    self.rateLimitTimer = nil
                }
            }
        }
    }

    // MARK: - Reactions

    func addReaction(to message: ChatMessage, emoji: String) async {
        do {
            try await apiClient.addReaction(messageId: message.id, emoji: emoji)
        } catch {
            // Silently handle reaction errors
        }
    }

    func removeReaction(from message: ChatMessage, emoji: String) async {
        do {
            try await apiClient.removeReaction(messageId: message.id, emoji: emoji)
        } catch {
            // Silently handle reaction errors
        }
    }

    // MARK: - Translation

    func translateMessage(_ message: ChatMessage) async -> String? {
        let langCode = Locale.current.language.languageCode?.identifier ?? "en"
        do {
            let result = try await apiClient.translateMessage(messageId: message.id, targetLanguage: langCode)
            return result.translatedText
        } catch {
            return nil
        }
    }

    // MARK: - Error

    func clearError() {
        error = nil
    }

    // MARK: - Socket.IO (wire in production polish pass)
    // In production: import SocketIO; replace stubs with SocketManager/SocketIOClient calls

    func connectSocket() {
        // TODO: production — SocketManager + auth token
    }

    func disconnectSocket() {
        // TODO: production — socket.disconnect()
    }
}

File: android/app/src/main/java/com/blakjaks/app/features/social/SocialViewModel.kt
package com.blakjaks.app.features.social

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.blakjaks.app.core.network.ApiClientInterface
import com.blakjaks.app.core.network.models.Channel
import com.blakjaks.app.core.network.models.ChatMessage
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

// ─── LiveStream model (defined here pending backend model promotion to Models.kt) ───

data class LiveStream(
    val id: Int,
    val title: String,
    val host: String,
    val isLive: Boolean,
    val viewerCount: Int,
    val streamUrl: String,
    val thumbnailUrl: String?
)

// ─── Mock default live stream ─────────────────────────────────────────────────

object MockLiveStream {
    val live = LiveStream(
        id = 1,
        title = "BlakJaks Live — New Flavor Drop",
        host = "BlakJaks Official",
        isLive = true,
        viewerCount = 842,
        streamUrl = "https://stream.blakjaks.com/live/hls/main.m3u8",
        thumbnailUrl = null
    )
}

// ─── SocialViewModel ──────────────────────────────────────────────────────────

class SocialViewModel(private val apiClient: ApiClientInterface) : ViewModel() {

    // ─── Channels ────────────────────────────────────────────────────────────

    private val _channels = MutableStateFlow<List<Channel>>(emptyList())
    val channels: StateFlow<List<Channel>> = _channels.asStateFlow()

    private val _selectedChannel = MutableStateFlow<Channel?>(null)
    val selectedChannel: StateFlow<Channel?> = _selectedChannel.asStateFlow()

    private val _isLoadingChannels = MutableStateFlow(false)
    val isLoadingChannels: StateFlow<Boolean> = _isLoadingChannels.asStateFlow()

    // ─── Messages ────────────────────────────────────────────────────────────

    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages.asStateFlow()

    private val _isLoadingMessages = MutableStateFlow(false)
    val isLoadingMessages: StateFlow<Boolean> = _isLoadingMessages.asStateFlow()

    // ─── Live Stream ─────────────────────────────────────────────────────────

    private val _currentLiveStream = MutableStateFlow<LiveStream?>(MockLiveStream.live)
    val currentLiveStream: StateFlow<LiveStream?> = _currentLiveStream.asStateFlow()

    // ─── Error ───────────────────────────────────────────────────────────────

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    // ─── Rate Limit ──────────────────────────────────────────────────────────

    private val _isRateLimited = MutableStateFlow(false)
    val isRateLimited: StateFlow<Boolean> = _isRateLimited.asStateFlow()

    private val _rateLimitRemainingSeconds = MutableStateFlow(0)
    val rateLimitRemainingSeconds: StateFlow<Int> = _rateLimitRemainingSeconds.asStateFlow()

    // ─── New Message Count / Pinned ───────────────────────────────────────────

    private val _newMessageCount = MutableStateFlow(0)
    val newMessageCount: StateFlow<Int> = _newMessageCount.asStateFlow()

    private val _pinnedMessage = MutableStateFlow<ChatMessage?>(null)
    val pinnedMessage: StateFlow<ChatMessage?> = _pinnedMessage.asStateFlow()

    // ─── Draft ───────────────────────────────────────────────────────────────

    val draftMessage = MutableStateFlow("")

    // ─── Private ─────────────────────────────────────────────────────────────

    private var lastSentAt: Long? = null
    private val rateLimitDuration = 5L // seconds
    private var rateLimitJob: Job? = null

    val currentUserId: Int = 1
    val currentUserTier: String = "Standard"

    // ─── Init ─────────────────────────────────────────────────────────────────

    init {
        loadChannels()
    }

    // ─── Channel Methods ──────────────────────────────────────────────────────

    fun loadChannels() {
        viewModelScope.launch {
            _isLoadingChannels.value = true
            try {
                _channels.value = apiClient.getChannels()
            } catch (e: Exception) {
                _error.value = e.message ?: "Failed to load channels"
            } finally {
                _isLoadingChannels.value = false
            }
        }
    }

    fun selectChannel(channel: Channel) {
        viewModelScope.launch {
            _selectedChannel.value = channel
            _messages.value = emptyList()
            _newMessageCount.value = 0
            loadMessages()
        }
    }

    fun loadMessages(before: Int? = null) {
        val channelId = _selectedChannel.value?.id ?: return
        viewModelScope.launch {
            _isLoadingMessages.value = true
            try {
                val fetched = apiClient.getMessages(channelId, 50, before)
                if (before != null) {
                    // Prepend older messages
                    _messages.value = fetched + _messages.value
                } else {
                    _messages.value = fetched
                }
            } catch (e: Exception) {
                _error.value = e.message ?: "Failed to load messages"
            } finally {
                _isLoadingMessages.value = false
            }
        }
    }

    // ─── Send Message ─────────────────────────────────────────────────────────

    fun sendMessage() {
        val trimmed = draftMessage.value.trim()
        if (trimmed.isEmpty()) return
        if (trimmed.length > 500) return
        if (_isRateLimited.value) return
        val channelId = _selectedChannel.value?.id ?: return

        viewModelScope.launch {
            try {
                val sent = apiClient.sendMessage(channelId, trimmed, null)
                _messages.value = _messages.value + sent
                draftMessage.value = ""
                lastSentAt = System.currentTimeMillis()
                if (currentUserTier.lowercase() == "standard") {
                    startRateLimitCooldown()
                }
            } catch (e: Exception) {
                _error.value = e.message ?: "Failed to send message"
            }
        }
    }

    // ─── Rate Limit Cooldown ──────────────────────────────────────────────────

    private fun startRateLimitCooldown() {
        rateLimitJob?.cancel()
        _isRateLimited.value = true
        _rateLimitRemainingSeconds.value = rateLimitDuration.toInt()

        rateLimitJob = viewModelScope.launch {
            repeat(rateLimitDuration.toInt()) {
                delay(1000L)
                _rateLimitRemainingSeconds.value--
            }
            _isRateLimited.value = false
        }
    }

    // ─── Reactions ────────────────────────────────────────────────────────────

    fun addReaction(message: ChatMessage, emoji: String) {
        viewModelScope.launch {
            try {
                apiClient.addReaction(message.id, emoji)
            } catch (_: Exception) {
                // Silently ignore reaction errors
            }
        }
    }

    fun removeReaction(message: ChatMessage, emoji: String) {
        viewModelScope.launch {
            try {
                apiClient.removeReaction(message.id, emoji)
            } catch (_: Exception) {
                // Silently ignore reaction errors
            }
        }
    }

    // ─── Translation ──────────────────────────────────────────────────────────

    suspend fun translateMessage(message: ChatMessage): String? {
        return try {
            val result = apiClient.translateMessage(message.id, "en")
            result.translatedText
        } catch (_: Exception) {
            null
        }
    }

    // ─── Socket.IO stubs (TODO: Wire to socket.io-client-java in production) ──

    fun connectSocket() {
        // TODO: Socket.IO — SocketManager + auth token handshake
    }

    fun disconnectSocket() {
        // TODO: Socket.IO — socket.disconnect()
    }

    // ─── Error ────────────────────────────────────────────────────────────────

    fun clearError() {
        _error.value = null
    }
}
